### PR TITLE
K8s: Add slog wrapper

### PR DESCRIPTION
**Pull request by toddtreece**
**What is this feature?**

This adds a `log/slog` wrapper to `pkg/infra/log`.

**Why do we need this feature?**

This will allow us to use the `slog` adapter that was recently added to `klog`.

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/infra/log/slog.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		h.Warn(r.Message, attrs...)
+	case slog.LevelError:
+		h.Error(r.Message, attrs...)
+	}
+	return nil
+}
+
+// WithAttrs implements Handler.WithAttrs.
+func (h *SLogHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
+	out := []any{}
```
Comment (charandas) [inline pkg/infra/log/slog.go:46]: You can likely preallocate with `2*len(attrs)`: similar to [here](https://github.com/kubernetes/klog/blob/main/internal/sloghandler/sloghandler_slog.go#L81)

Comment (toddtreece) [reply pkg/infra/log/slog.go]: good call. updated

File: pkg/infra/log/slog.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+import (
+	"context"
+	"log/slog"
+)
+
+var _ slog.Handler = &SLogHandler{}
+
+type SLogHandler struct {
+	Logger
```
Comment (marefr) [inline pkg/infra/log/slog.go:11]: Nit does this need to be exported?

File: pkg/infra/log/slog.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+// NewSLogHandler returns a SLogHandler with the given level.
+func NewSLogHandler(logger Logger) *SLogHandler {
```
Comment (marefr) [inline pkg/infra/log/slog.go:15]: Maybe in a child package instead, slogadapter, with `New`?

File: pkg/services/apiserver/options/extra.go [hunk 4]
```diff
@@ -1,9 +1,10 @@
 package options
 
 import (
+	"log/slog"
 	"strconv"
 
-	"github.com/go-logr/logr"
```
Comment (marefr) [inline pkg/services/apiserver/options/extra.go:6]: This still in use or go mod tidy removes it?

File: pkg/infra/log/slog.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	}
+	return NewSLogHandler(h.Logger.New(out...))
+}
+
+// WithGroup implements Handler.WithGroup.
+func (h *SLogHandler) WithGroup(name string) slog.Handler {
+	if name == "" {
+		return h
+	}
+	return NewSLogHandler(h.Logger.New(name))
```
Comment (marefr) [inline pkg/infra/log/slog.go:59]: I might be wrong, but doesn't Logger.New require a key/value pair?

**Review by marefr — COMMENTED** [review]
Looks good, but some comments/suggestions

File: pkg/infra/log/slog.go (hunk 5 above)
Comment (toddtreece) [reply pkg/infra/log/slog.go]: added `"group"` key in [de75497](https://github.com/grafana/grafana/pull/84680/commits/de75497dc93d208bd0ef0d7e26b689aeffb71a50)

File: pkg/services/apiserver/options/extra.go (hunk 4 above)
Comment (toddtreece) [reply pkg/services/apiserver/options/extra.go:6]: good catch. it's no longer a direct dependency. updated `go.mod` in [bb15b87](https://github.com/grafana/grafana/pull/84680/commits/bb15b876131e70e607b68eec391aa45497d16be8)

File: pkg/infra/log/slog.go (hunk 3 above)
Comment (toddtreece) [reply pkg/infra/log/slog.go]: updated in [de75497](https://github.com/grafana/grafana/pull/84680/commits/de75497dc93d208bd0ef0d7e26b689aeffb71a50)

File: pkg/infra/log/slog.go (hunk 2 above)
Comment (toddtreece) [reply pkg/infra/log/slog.go]: unexported in [de75497](https://github.com/grafana/grafana/pull/84680/commits/de75497dc93d208bd0ef0d7e26b689aeffb71a50)

**Review by kylebrandt — APPROVED** [review]

File: pkg/infra/log/slogadapter/adapter.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+)
+
+var _ slog.Handler = &slogHandler{}
+
+type slogHandler struct {
+	log.Logger
+}
+
+// NewSLogHandler returns a new slog.Handler that logs to the given log.Logger.
+func NewSLogHandler(logger log.Logger) *slogHandler {
```
Comment (marefr) [inline pkg/infra/log/slogadapter/adapter.go:17]:
```suggestion
func New(logger log.Logger) *slogHandler {
```

File: pkg/services/apiserver/options/extra.go [hunk 7]
```diff
@@ -36,10 +38,12 @@ func (o *ExtraOptions) Validate() []error {
 }
 
 func (o *ExtraOptions) ApplyTo(c *genericapiserver.RecommendedConfig) error {
-	logger := logr.New(newLogAdapter(o.Verbosity))
-	klog.SetLoggerWithOptions(logger, klog.ContextualLogger(true))
+	handler := slogadapter.NewSLogHandler(log.New("grafana-apiserver"))
```
Comment (marefr) [inline pkg/services/apiserver/options/extra.go:41]:
```suggestion
	handler := slogadapter.New(log.New("grafana-apiserver"))
```

**Review by marefr — APPROVED** [review]
LGTM with a suggestion that I think reads better.

**Review by charandas — APPROVED** [review]

**Pull request merged**